src/backup.py

```diff
@@ -128,3 +161,255 @@ def can_use_bucket(self) -> bool:
+
+    Raises:
+       BackupError: if there was an error preparing the configuration.
+    """
+    aws_set_addressing_style = [
+        AWS_COMMAND,
+        "configure",
+        "set",
+        "default.s3.addressing_style",
+        s3_parameters.addressing_style,
+    ]
+
+    aws_set_concurrent_requests = [
+        AWS_COMMAND,
+        "configure",
+        "set",
+        "default.s3.max_concurrent_requests",
+        str(S3_MAX_CONCURRENT_REQUESTS),
+    ]
+
+    try:
+        process = container.exec(
+            aws_set_addressing_style,
+            user=synapse.SYNAPSE_USER,
+            group=synapse.SYNAPSE_GROUP,
+        )
+        process.wait()
+        process = container.exec(
+            aws_set_concurrent_requests,
+            user=synapse.SYNAPSE_USER,
+            group=synapse.SYNAPSE_GROUP,
+        )
+        process.wait()
+    except (APIError, ExecError) as exc:
+        raise BackupError("Backup Failed. Error configuring AWS.") from exc
+
+    try:
+        container.push(
+            PASSPHRASE_FILE,
+            passphrase,
+            user=synapse.SYNAPSE_USER,
+            group=synapse.SYNAPSE_GROUP,
+        )
+    except ops.pebble.PathError as exc:
+        raise BackupError("Backup Failed. Error configuring GPG passphrase.") from exc
+
+
+def _get_paths_to_backup(container: ops.Container) -> Iterable[str]:
+    """Get the list of paths that should be in a backup for Synapse.
+
+    Args:
+       container: Synapse Container.
+
+    Returns:
+       Iterable with the list of paths to backup.
+    """
+    paths = []
+    for pattern in BACKUP_FILE_PATTERNS:
+        paths += container.list_files(synapse.SYNAPSE_CONFIG_DIR, pattern=pattern)
+    # Local media if it exists
+    media_dir = os.path.join(synapse.SYNAPSE_DATA_DIR, MEDIA_DIR)
+    if container.exists(media_dir):
+        paths += container.list_files(media_dir, pattern=LOCAL_DIR_PATTERN)
+    return [path.path for path in paths]
+
+
+def _calculate_size(container: ops.Container, paths: Iterable[str]) -> int:
+    """Return the combined size of all the paths given.
+
+    Args:
+        container: Container where to check the size of the paths.
+        paths: Paths to check.
+
+    Returns:
+        Total size in bytes.
+
+    Raises:
+        BackupError: If there was a problem calculating the size.
+    """
+    command = "set -euxo pipefail; du -bsc " + _paths_to_args(paths) + " | tail -n1 | cut -f 1"
+    try:
+        exec_process = container.exec(
+            [BASH_COMMAND, "-c", command],
+            user=synapse.SYNAPSE_USER,
+            group=synapse.SYNAPSE_GROUP,
+        )
+        stdout, stderr = exec_process.wait_output()
+    except (APIError, ExecError) as exc:
+        raise BackupError("Cannot calculate size of paths. du failed.") from exc
+
+    logger.info(
+        "Calculating size of paths. Command: %s. stdout: %s. stderr: %s", command, stdout, stderr
+    )
+
+    try:
+        return int(stdout)
```

Comment:
Is this even possible considering we should have received a 0 exit code?

Reply:
I understand this should not happen...

Reply:
I think it would make sense to directly let the charm error if this happens

Reply:
Removed the try catch that should not happen!